libraries/AC_AutoTune/AC_AutoTune_Multi.cpp

```diff
@@ -0,0 +1,626 @@
+        break;
+    }
+}
+
+// save_tuning_gains - save the final tuned gains for each axis
+// save discovered gains to eeprom if autotuner is enabled (i.e. switch is in the high position)
+void AC_AutoTune_Multi::save_tuning_gains()
+{
+
+    AC_AutoTune::save_tuning_gains();
+
+    // sanity check the rate P values
+    if ((axes_completed & AUTOTUNE_AXIS_BITMASK_ROLL) && roll_enabled() && !is_zero(tune_roll_rp)) {
+        // rate roll gains
+        attitude_control->get_rate_roll_pid().ff(orig_roll_rff);
+        attitude_control->get_rate_roll_pid().filt_T_hz(orig_roll_fltt);
+        attitude_control->get_rate_roll_pid().kI(tune_roll_rp*AUTOTUNE_PI_RATIO_FINAL);
+        attitude_control->get_rate_roll_pid().save_gains();
+
+        // resave pids to originals in case the autotune is run again
+        orig_roll_rff = attitude_control->get_rate_roll_pid().ff();
+        orig_roll_ri = attitude_control->get_rate_roll_pid().kI();
+    }
+
+    if ((axes_completed & AUTOTUNE_AXIS_BITMASK_PITCH) && pitch_enabled() && !is_zero(tune_pitch_rp)) {
+        // rate pitch gains
+        attitude_control->get_rate_pitch_pid().ff(orig_pitch_rff);
+        attitude_control->get_rate_pitch_pid().filt_T_hz(orig_pitch_fltt);
+        attitude_control->get_rate_pitch_pid().kI(tune_pitch_rp*AUTOTUNE_PI_RATIO_FINAL);
+        attitude_control->get_rate_pitch_pid().save_gains();
+
+        // resave pids to originals in case the autotune is run again
+        orig_pitch_rff = attitude_control->get_rate_pitch_pid().ff();
+        orig_pitch_ri = attitude_control->get_rate_pitch_pid().kI();
+    }
+
+    if ((axes_completed & AUTOTUNE_AXIS_BITMASK_YAW) && yaw_enabled() && !is_zero(tune_yaw_rp)) {
+        // rate yaw gains
+        attitude_control->get_rate_yaw_pid().kD(0.0f);
+        attitude_control->get_rate_yaw_pid().ff(orig_yaw_rff);
+        attitude_control->get_rate_yaw_pid().filt_T_hz(orig_yaw_fltt);
+        attitude_control->get_rate_yaw_pid().filt_E_hz(tune_yaw_rLPF);
+        attitude_control->get_rate_yaw_pid().kI(tune_yaw_rp*AUTOTUNE_YAW_PI_RATIO_FINAL);
+        attitude_control->get_rate_yaw_pid().save_gains();
+
+        // resave pids to originals in case the autotune is run again
+        orig_yaw_rd = attitude_control->get_rate_yaw_pid().kD();
+        orig_yaw_rLPF = attitude_control->get_rate_yaw_pid().filt_E_hz();
+        orig_yaw_rff = attitude_control->get_rate_yaw_pid().ff();
+        orig_yaw_ri = attitude_control->get_rate_yaw_pid().kI();
+    }
+
+    // update GCS and log save gains event
+    update_gcs(AUTOTUNE_MESSAGE_SAVED_GAINS);
+    AP::logger().Write_Event(LogEvent::AUTOTUNE_SAVEDGAINS);
+
+    reset();
+}
+
+void AC_AutoTune_Multi::updating_rate_p_up_all(uint8_t test_axis)
+{
+    switch (test_axis) {
+    case ROLL:
+        updating_rate_p_up_d_down(tune_roll_rd, min_d, AUTOTUNE_RD_STEP, tune_roll_rp, AUTOTUNE_RP_MIN, AUTOTUNE_RP_MAX, AUTOTUNE_RP_STEP, target_rate, test_rate_min, test_rate_max);
```

Comment:
not included in this PR but is there a typo above?  It seems like updating_rate_**p**_up_all is calling updating_rate_p_up_**d**_down ("p" vs "d")

Reply:
@rmackay9  I was having a tough time when trying to break them into sub classes because of some unique methods used in the multi autotune.  I'm open to ideas for a better way to do this.  I chose to make the tests generic that any frame could choose the test.  the updating gain methods did not fit into a nice construct.  So I made the construct of Rate_P_Up, Rate_P_Down...   I just categorized the updating_rate_p_up_d_down in one of the bins.  The flow follows the same flow as the multicopter autotune, it just isn't named the same because I was trying to put a round peg into square hole.  Open to suggestions.